java/arcs/core/storage/api/EntityBase.kt

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.storage.api
+
+import arcs.core.common.Id
+import arcs.core.common.Referencable
+import arcs.core.data.FieldType
+import arcs.core.data.PrimitiveType
+import arcs.core.data.RawEntity
+import arcs.core.data.RawEntity.Companion.NO_REFERENCE_ID
+import arcs.core.data.Schema
+import arcs.core.data.util.ReferencablePrimitive
+import arcs.core.data.util.toReferencable
+
+open class EntityBase(
+    private val entityClassName: String,
+    private val schema: Schema
+) : Entity {
+    // Private var _entityId with public getter. Only this class can set this field.
+    private var _entityId: String? = null
+    override val entityId: String?
+        get() = _entityId
+
+    private val singletons: MutableMap<String, Any?> = mutableMapOf()
+    private val collections: MutableMap<String, Set<Any>> = mutableMapOf()
+
+    // Initialize all fields. After this point, if a key is not present in singletons/collections,
+    // it will not be considered a valid field for the entity.
+    init {
+        reset()
+    }
+
+    /** Returns the value for the given singleton field. */
+    protected fun getSingletonValue(field: String): Any? = if (field in singletons) {
+        singletons[field]
```

Comment:
getOrElse?

Reply:
Unfortunately getOrElse throws if the element is null (also, it throws an UnsupportedOperationException, and I wanted to keep the error messages the same for when you call set or get)